rs/execution_environment/src/scheduler/round_schedule.rs

```diff
@@ -229,4 +231,77 @@ impl RoundSchedule {
 
         (canisters_partitioned_by_cores, canisters)
     }
+
+    /// Finishes canister execution on thread.
+    pub fn finish_canister_execution_on_thread(
+        canister: &mut CanisterState,
+        executed_priority_ids: &mut BTreeMap<CanisterId, AccumulatedPriority>,
+        round_id: ExecutionRound,
+        is_first_iteration: bool,
+        rank: usize,
+    ) {
+        let full_message_execution = match canister.next_execution() {
+            NextExecution::None => true,
+            NextExecution::StartNew => false,
+            // We just finished a full slice of executions.
+            NextExecution::ContinueLong | NextExecution::ContinueInstallCode => true,
+        };
+        let scheduled_first = is_first_iteration && rank == 0;
+
+        if full_message_execution || scheduled_first {
+            // The very first canister is considered to have a full execution round for
+            // scheduling purposes even if it did not complete within the round.
+            canister.scheduler_state.last_full_execution_round = round_id;
+
+            // For now the priority for full execution is always 100.
+            // In future we might use a proportion to the actual load.
+            executed_priority_ids.insert(canister.canister_id(), 100.into());
+        }
+    }
+
+    /// Finishes inner round execution.
+    pub(crate) fn finish_inner_round(
+        &self,
+        canister_states: &mut BTreeMap<CanisterId, CanisterState>,
+        executed_priority_ids: BTreeMap<CanisterId, AccumulatedPriority>,
+    ) {
+        let scheduler_cores = self.scheduler_cores;
+        let compute_capacity_percent =
+            SchedulerImpl::compute_capacity_percent(scheduler_cores) as i64;
+        let number_of_canisters = canister_states.len();
+        let multiplier = (scheduler_cores * number_of_canisters).max(1) as i64;
+
+        // Charge canisters for the priority executed in the previous round.
+        let mut total_executed_priority = 0;
+        for (canister_id, executed_priority) in executed_priority_ids {
+            if let Some(canister) = canister_states.get_mut(&canister_id) {
+                total_executed_priority += executed_priority.get() * multiplier;
+                // If priority credit is non-zero, use it instead of accumulated priority.
+                if canister.scheduler_state.priority_credit != 0.into() {
+                    canister.scheduler_state.priority_credit += executed_priority * multiplier;
+                } else {
+                    canister.scheduler_state.accumulated_priority -= executed_priority * multiplier;
+                }
```

Comment:
I have to admit that I'm a bit lost, but aren't we still charging the first 4 canisters somewhere? I don't see anywhere in the PR diff the equivalent code being removed.

Reply:
That's a "shortcut". We charge top canisters and on top for charged priority.

Reply:
This will be solved when we also charge idle canisters as fully executed, which is not part of this PR.